File: src/plcc/lang/ext/python/runtime/deserialize.py
from .base import Token


def deserialize(tree, registry):
    if tree['kind'] == 'token':
        return Token(kind=tree['name'], lexeme=tree['lexeme'])

    children = tree.get('children', [])
    field_names = [pair[0] for pair in children]
    cls = registry.lookup(tree['rule'], field_names)
    kwargs = {name: _deserialize_value(val, registry) for name, val in children}
    return cls(**kwargs)


def _deserialize_value(val, registry):
    if isinstance(val, list):
        return [_deserialize_value(item, registry) for item in val]
    if isinstance(val, dict) and val.get('kind') == 'token':
        return Token(kind=val['name'], lexeme=val['lexeme'])
    if isinstance(val, dict):
        return deserialize(val, registry)
    return val

File: src/plcc/lang/ext/javascript/runtime/deserialize.js
const { Token } = require('./base');

function deserialize(tree, registry) {
    if (tree.kind === 'token') {
        return new Token(tree.name, tree.lexeme);
    }
    const children = tree.children || [];
    const fieldNames = children.map(([name]) => name);
    const cls = registry.lookup(tree.rule, fieldNames);
    const args = children.map(([, val]) => _deserializeValue(val, registry));
    return new cls(...args);
}

function _deserializeValue(val, registry) {
    if (Array.isArray(val)) {
        return val.map(item => _deserializeValue(item, registry));
    }
    if (val && typeof val === 'object' && val.kind === 'token') {
        return new Token(val.name, val.lexeme);
    }
    if (val && typeof val === 'object') {
        return deserialize(val, registry);
    }
    return val;
}

module.exports = { deserialize };
